osu.Game.Tests/Visual/Online/TestSceneCommentsContainer.cs

```diff
@@ -30,30 +34,117 @@ public class TestSceneCommentsContainer : OsuTestScene
 
         public TestSceneCommentsContainer()
         {
-            BasicScrollContainer scrollFlow;
+            CommentsContainer commentsContainer = new CommentsContainer();
+            BasicScrollContainer scroll;
 
-            Add(scrollFlow = new BasicScrollContainer
+            Add(scroll = new BasicScrollContainer
             {
                 RelativeSizeAxes = Axes.Both,
+                Child = commentsContainer
             });
 
-            AddStep("Big Black comments", () =>
+            AddStep("Idle state", () =>
             {
-                scrollFlow.Clear();
-                scrollFlow.Add(new CommentsContainer(CommentableType.Beatmapset, 41823));
+                scroll.Clear();
+                scroll.Add(commentsContainer = new CommentsContainer());
             });
+            AddStep("Big Black comments", () => commentsContainer.ShowComments(CommentableType.Beatmapset, 41823));
+            AddStep("Airman comments", () => commentsContainer.ShowComments(CommentableType.Beatmapset, 24313));
+            AddStep("lazer build comments", () => commentsContainer.ShowComments(CommentableType.Build, 4772));
+            AddStep("local comments", () => commentsContainer.ShowComments(comment_bundle));
+            AddStep("local empty comments", () => commentsContainer.ShowComments(empty_comment_bundle));
+        }
 
-            AddStep("Airman comments", () =>
-            {
-                scrollFlow.Clear();
-                scrollFlow.Add(new CommentsContainer(CommentableType.Beatmapset, 24313));
-            });
+        [BackgroundDependencyLoader]
+        private void load(IAPIProvider api)
+        {
+            AddStep("trigger logout", api.Logout);
```

Comment:
You're using the actual online API, and adding a step to logout to check comment availability when logged out. However, due to the fact you're using the actual online API, you actually *can't log back in again in this same test*. Moreover, *this silently logs you out from the actual lazer client*. This level of state pollution is unacceptable in tests as far as I'm concerned.

What this basically means is this is too tightly coupled to online components now to be testable, which is not good. [There were warnings about that too, by the way.](https://github.com/ppy/osu/pull/6504#issuecomment-542063407)

Reply:
I just splitted the test scene into two: online and offline ones. In online you can just check whether all the stuff is working and how it looks and for offline you can test your local bundles